src/track/globaltrackcache.h

```diff
@@ -208,37 +157,62 @@ class GlobalTrackCache {
     // to verify the class invariants during development.
     bool verifyConsistency() const;
 
-    TrackPointer lookupInternal(
-            const TrackId& trackId) const;
-    TrackPointer lookupInternal(
-            const TrackRef& trackRef) const;
+    void reset();
 
-    bool resolveInternal(
-            GlobalTrackCacheResolver* pCacheResolver,
-            TrackRef* pTrackRef,
-            const TrackId& trackId,
-            const QFileInfo& fileInfo);
+    TrackPointer lookupById(
+            const TrackId& trackId);
+    TrackPointer lookupByRef(
+            const TrackRef& trackRef);
 
-    TrackRef updateTrackIdInternal(
-            const TrackPointer& pTrack,
-            const TrackRef& trackRef,
+    TrackPointer revive(
+            Track* plainPtr);
+
+    void resolve(
+            GlobalTrackCacheResolver* /*in/out*/ pCacheResolver,
+            QFileInfo /*in*/ fileInfo,
+            TrackId /*in*/ trackId,
+            SecurityTokenPointer /*in*/ pSecurityToken);
+
+    TrackRef initTrackId(
+            const TrackPointer& strongPtr,
+            TrackRef trackRef,
             TrackId trackId);
 
-    Item purgeInternal(
-            const TrackRef& trackRef);
+    bool evictAndDelete(
+            Track* plainPtr);
+
+    typedef std::unordered_map<Track*, TrackWeakPointer> IndexedTracks;
 
-    void evict(
-            Track* pTrack);
-    Track* evictInternal(
+    bool evictAndDelete(
             GlobalTrackCacheLocker* /*nullable*/ pCacheLocker,
-            const TrackRef& trackRef);
+            IndexedTracks::iterator indexedTrack,
+            bool evictUnexpired);
+    bool evict(
+            const TrackRef& trackRef,
+            IndexedTracks::iterator indexedTrack,
+            bool evictUnexpired);
 
-    GlobalTrackCacheEvictor* m_pEvictor;
+    void afterEvicted(
+            GlobalTrackCacheLocker* /*nullable*/ pCacheLocker,
+            Track* plainPtr);
+
+    bool isEmpty() const;
 
+    void deactivate();
+
+    // Managed by GlobalTrackCacheLocker
     mutable QMutex m_mutex;
 
-    typedef QHash<TrackId, Item> TracksById;
+    GlobalTrackCacheEvictor* m_pEvictor;
+
+    IndexedTracks m_indexedTracks;
+
+    typedef std::unordered_set<Track*> UnindexedTracks;
+    UnindexedTracks m_unindexedTracks;
```

Comment:
where are tracks added to this container?

Reply:
This set stores the set of track objects that were actively managed while the cache was reset() (= disabled) during shutdown. Those objects need a special case handling when their deleter kicks in since they won't be found in the (empty) cache. Just to keep the regular logic strict and avoid any special case handling on the regular path.